apps/potterhat_node/lib/node/node.ex

```diff
@@ -0,0 +1,261 @@
+        Process.sleep(retry_period_ms)
+        {:noreply, %{state | state: :started}, {:continue, :print_version}}
+    end
+  end
+
+  @impl true
+  def handle_continue(:register_with_manager, state) do
+    _ =
+      case state.node_registry do
+        nil -> :noop
+        registry -> registry.register(self(), state.priority)
+      end
+
+    {:noreply, state, {:continue, :subscribe_events}}
+  end
+
+  @impl true
+  def handle_continue(:subscribe_events, state) do
+    opts = [
+      node_id: state[:id],
+      label: state[:label],
+      listener: self()
+    ]
+
+    listening_events =
+      :potterhat_node
+      |> Application.get_env(:listen_events)
+      |> Enum.reduce([], fn
+        :sync_status, events ->
+          SyncStatus.start_link(state[:ws], opts)
+          [:sync_status | events]
+
+        :new_head, events ->
+          NewHead.start_link(state[:ws], opts)
+          [:new_head | events]
+
+        :new_pending_transaction, events ->
+          NewPendingTransaction.start_link(state[:ws], opts)
+          [:new_pending_transaction | events]
+
+        :log, events ->
+          Log.start_link(state[:ws], opts)
+          [:log | events]
+
+        _, events ->
+          events
+      end)
+
+    {:noreply, %{state | state: :listening, listening_events: listening_events}}
+  end
+
+  @impl true
+  def handle_call(:get_priority, _from, state) do
+    {:reply, state.priority, state}
+  end
+
+  @impl true
+  def handle_call(:get_label, _from, state) do
+    {:reply, state.label, state}
+  end
+
+  @impl true
+  def handle_call({:subscribe, subscriber}, _from, state) do
+    {response, new_state} =
+      state.subscribers
+      |> Enum.member?(subscriber)
+      |> case do
+        false ->
+          subscribed = [subscriber | state.subscribers]
+          {:ok, %{state | subscribers: subscribed}}
+
+        true ->
+          {{:error, :already_subscribed}, state}
+      end
+
+    {:reply, response, new_state}
+  end
+
+  @impl true
+  def handle_call({:unsubscribe, unsubscriber}, _from, state) do
+    {response, new_state} =
+      state.subscribers
+      |> Enum.member?(unsubscriber)
+      |> case do
+        true ->
+          unsubscribed = List.delete(state.subscribers, unsubscriber)
+          {:ok, %{state | subscribers: unsubscribed}}
+
+        false ->
+          {{:error, :not_subscribed}, state}
+      end
+
+    {:reply, response, new_state}
+  end
+
+  @impl true
+  def handle_call(:get_subscribers, _from, state) do
+    {:reply, state.subscribers, state}
+  end
+
+  @impl true
+  def handle_call({:rpc_request, body_params, header_params}, _from, state) do
+    encoded_params = Jason.encode!(body_params)
+
+    # Send only supported headers. Infura doesn't like extra headers.
+    header_params = Enum.filter(header_params, fn
+      {"content-type", _} -> true
+      _ -> false
+    end)
+
+    _ = HTTPoison.start()
```

Comment:
hm, what does this do?

Reply:
Good question. This starts up the HTTPoison client, but it should be started as part of the application startup, not here.